File: src/main/java/eu/timerertim/downlomatic/utils/JSoupDriver.java
package eu.timerertim.downlomatic.utils;

import org.jsoup.Connection;
import org.jsoup.helper.HttpConnection;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.*;
import org.openqa.selenium.internal.*;
import org.openqa.selenium.logging.Logs;

import java.io.IOException;
import java.net.URL;
import java.util.*;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.stream.Collectors;

/**
 * JSoup based implementation of Selenium's {@link WebDriver} for uniform webpage parsing procedure.
 */
public class JSoupDriver implements WebDriver, SearchContext {
    private final static String USER_AGENT = "Mozilla/5.0 (Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0";

    private final Map<String, String> cookies;
    private Options options;
    private Connection con;
    private Document doc;
    private URL url;

    public JSoupDriver() {
        cookies = new HashMap<>();
        generateOptions();
    }

    @Override
    public void get(String url) {
        get(url, (con) -> {
        });
    }

    public void get(String url, Consumer<Connection> dataModifier) {
        generateConnection();
        dataModifier.accept(con);
        try {
            Connection.Response res = con.url(url).cookies(cookies).execute();
            cookies.putAll(res.cookies());
            doc = res.parse();
            this.url = new URL(url);
        } catch (IOException e) {
            java.util.logging.Logger.getLogger("org.openqa.selenium").log(Level.WARNING, getClass().getSimpleName() + " was not able to parse from URL: " + url, e);
            doc = null;
            this.url = null;
        } catch (NullPointerException e) {
            java.util.logging.Logger.getLogger("org.openqa.selenium").log(Level.SEVERE, getClass().getSimpleName() + " is closed", e);
            doc = null;
            this.url = null;
        }
    }

    @Override
    public String getCurrentUrl() {
        if (url != null) {
            return url.toString();
        } else
            return null;
    }

    @Override
    public String getTitle() {
        if (doc != null) {
            return doc.title();
        } else
            return null;
    }

    @Override
    public List<WebElement> findElements(By by) {
        return by.findElements(new JSoupWebElement(doc));
    }

    @Override
    public WebElement findElement(By by) {
        return by.findElement(new JSoupWebElement(doc));
    }

    @Override
    public String getPageSource() {
        if (doc != null) {
            return doc.toString();
        } else
            return null;
    }

    @Override
    public void close() {
        url = null;
        doc = null;
        con = null;
    }

    @Override
    public void quit() {
        close();
        options = null;
    }

    @Override
    public Set<String> getWindowHandles() {
        return null;
    }

    @Override
    public String getWindowHandle() {
        return null;
    }

    @Override
    public TargetLocator switchTo() {
        return null;
    }

    @Override
    public Navigation navigate() {
        return null;
    }

    @Override
    public Options manage() {
        return options;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + ": JSoup on " + System.getProperty("os.name") + " (" + options + ")";
    }

    private void generateConnection() {
        con = new HttpConnection().userAgent(USER_AGENT);
    }

    private void generateOptions() {
        options = new Options() {
            @Override
            public void addCookie(Cookie cookie) {
                cookies.put(cookie.getName(), cookie.getValue());
            }

            @Override
            public void deleteCookieNamed(String name) {
                cookies.remove(name);
            }

            @Override
            public void deleteCookie(Cookie cookie) {
                cookies.remove(cookie.getName(), cookie.getValue());
            }

            @Override
            public void deleteAllCookies() {
                cookies.clear();
            }

            @Override
            public Set<Cookie> getCookies() {
                return cookies.entrySet().stream().map((cookieEntry) -> new Cookie(cookieEntry.getKey(), cookieEntry.getValue())).collect(Collectors.toSet());
            }

            @Override
            public Cookie getCookieNamed(String name) {
                String value = cookies.get(name);
                return (value == null ? null : new Cookie(name, value));
            }

            @Override
            public Timeouts timeouts() {
                return null;
            }

            @Override
            public ImeHandler ime() {
                return null;
            }

            @Override
            public Window window() {
                return null;
            }

            @Override
            public Logs logs() {
                return null;
            }
        };
    }

    private static class JSoupWebElement implements WebElement, FindsByCssSelector, FindsByXPath, FindsById, FindsByName, FindsByClassName, FindsByTagName, FindsByLinkText {
        // All Bys for JSoupWebElements
        private static final EditableBy cssSelector = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.getAllElements().stream().
                        filter(element -> element.cssSelector().equals(selector)).
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy xPath = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return null;
            }
        };
        private static final EditableBy id = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsByAttributeValueContaining("id", selector).stream().
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy name = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsByAttributeValue("name", selector).stream().
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy className = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsByClass(selector).stream().
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy tagName = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsByTag(selector).stream().
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy linkText = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsMatchingText(selector).stream().filter(element -> element.tagName().equals("a")).
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private static final EditableBy partialLinkText = new EditableBy() {
            public List<WebElement> findElements(SearchContext context) {
                return ((JSoupWebElement) context).element.
                        getElementsContainingText(selector).stream().filter(element -> element.tagName().equals("a")).
                        map(JSoupWebElement::new).
                        collect(Collectors.toCollection(LinkedList::new));
            }
        };
        private final Element element;

        public JSoupWebElement(Element element) {
            this.element = element;
        }

        @Override
        public void click() {
        }

        @Override
        public void submit() {
        }

        @Override
        public void sendKeys(CharSequence... keysToSend) {
        }

        @Override
        public void clear() {
            element.val("");
        }

        @Override
        public String getTagName() {
            return element.tagName();
        }

        @Override
        public String getAttribute(String name) {
            return element.attr(name);
        }

        @Override
        public boolean isSelected() {
            return false;
        }

        @Override
        public boolean isEnabled() {
            return true;
        }

        @Override
        public String getText() {
            return element.text();
        }

        @Override
        public List<WebElement> findElements(By by) {
            return by.findElements(this);
        }

        @Override
        public WebElement findElement(By by) {
            return by.findElement(this);
        }

        @Override
        public boolean isDisplayed() {
            return false;
        }

        @Override
        public Point getLocation() {
            return null;
        }

        @Override
        public Dimension getSize() {
            return null;
        }

        @Override
        public Rectangle getRect() {
            return null;
        }

        @Override
        public String getCssValue(String propertyName) {
            return "";
        }

        @Override
        public <X> X getScreenshotAs(OutputType<X> target) throws WebDriverException {
            return null;
        }

        @Override
        public WebElement findElementByClassName(String using) {
            return className.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByClassName(String using) {
            return className.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementById(String using) {
            return id.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsById(String using) {
            return id.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByLinkText(String using) {
            return linkText.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByLinkText(String using) {
            return linkText.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByPartialLinkText(String using) {
            return partialLinkText.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByPartialLinkText(String using) {
            return partialLinkText.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByName(String using) {
            return name.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByName(String using) {
            return name.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByTagName(String using) {
            return tagName.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByTagName(String using) {
            return tagName.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByCssSelector(String using) {
            return cssSelector.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByCssSelector(String using) {
            return cssSelector.setSelector(using).findElements(this);
        }

        @Override
        public WebElement findElementByXPath(String using) {
            return xPath.setSelector(using).findElement(this);
        }

        @Override
        public List<WebElement> findElementsByXPath(String using) {
            return xPath.setSelector(using).findElements(this);
        }

        private Element getElement() {
            return element;
        }

        private abstract static class EditableBy extends By {
            protected String selector;

            public By setSelector(String selector) {
                this.selector = selector;
                return this;
            }

            @Override
            public WebElement findElement(SearchContext context) {
                try {
                    return super.findElement(context);
                } catch (NoSuchElementException ex) {
                    return null;
                }
            }
        }
    }
}

File: server/src/main/kotlin/eu/timerertim/downlomatic/utils/JSoupDriver.kt
package eu.timerertim.downlomatic.utils

import eu.timerertim.downlomatic.utils.logging.Log.f
import eu.timerertim.downlomatic.utils.logging.Log.w
import org.jsoup.Connection
import org.jsoup.helper.HttpConnection
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.openqa.selenium.*
import org.openqa.selenium.NoSuchElementException
import org.openqa.selenium.WebDriver.*
import org.openqa.selenium.internal.*
import org.openqa.selenium.logging.Logs
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.stream.Collectors

/**
 * JSoup based implementation of Selenium's [WebDriver] for uniform webpage parsing procedure.
 */
class JSoupDriver : WebDriver, SearchContext {
    private val cookies: MutableMap<String, String>
    private var options: Options? = null
    private var con: Connection? = null
    private var doc: Document? = null
    private var url: URL? = null

    /**
     * Returns the response status message.
     *
     * @return the status message
     */
    var statusMessage: String? = null
        private set

    override operator fun get(url: String) {
        get(url) {}
    }

    operator fun get(url: String, dataModifier: Connection?.() -> Unit) {
        generateConnection()
        dataModifier(con)
        try {
            val res = con!!.url(url).cookies(cookies).execute()
            statusMessage = res.statusMessage()
            cookies.putAll(res.cookies())
            doc = res.parse()
            this.url = URL(url)
        } catch (e: IOException) {
            w(javaClass.simpleName + " was not able to parse from URL: " + url, e)
            doc = null
            statusMessage = null
            this.url = null
        } catch (e: NullPointerException) {
            f(javaClass.simpleName + " is closed", e)
            doc = null
            statusMessage = null
            this.url = null
        }
    }

    override fun getCurrentUrl() = url?.toString()

    override fun getTitle() = doc?.title()

    override fun findElements(by: By): List<WebElement> {
        return if (doc == null) LinkedList() else by.findElements(JSoupWebElement(doc!!))
    }

    override fun findElement(by: By): WebElement? {
        return if (doc == null) null else by.findElement(JSoupWebElement(doc!!))
    }

    override fun getPageSource() = doc?.toString()

    override fun close() {
        url = null
        doc = null
        con = null
    }

    override fun quit() {
        close()
        options = null
    }

    override fun getWindowHandles(): Set<String>? {
        return null
    }

    override fun getWindowHandle(): String? {
        return null
    }

    override fun switchTo(): TargetLocator? {
        return null
    }

    override fun navigate(): Navigation? {
        return null
    }

    override fun manage(): Options? {
        return options
    }

    override fun toString(): String {
        return javaClass.simpleName + ": JSoup on " + System.getProperty("os.name") + " (" + options + ")"
    }

    private fun generateConnection() {
        con = HttpConnection().userAgent(USER_AGENT).ignoreHttpErrors(true)
    }

    private fun generateOptions() {
        options = object : Options {
            override fun addCookie(cookie: Cookie) {
                this@JSoupDriver.cookies[cookie.name] = cookie.value
            }

            override fun deleteCookieNamed(name: String) {
                this@JSoupDriver.cookies.remove(name)
            }

            override fun deleteCookie(cookie: Cookie) {
                this@JSoupDriver.cookies.remove(cookie.name, cookie.value)
            }

            override fun deleteAllCookies() {
                this@JSoupDriver.cookies.clear()
            }

            override fun getCookies(): Set<Cookie> {
                return this@JSoupDriver.cookies.entries.stream().map { (key, value) ->
                    Cookie(
                        key, value
                    )
                }.collect(Collectors.toSet())
            }

            override fun getCookieNamed(name: String): Cookie? {
                val value: String? = this@JSoupDriver.cookies[name]
                return if (value == null) null else Cookie(name, value)
            }

            override fun timeouts(): Timeouts? {
                return null
            }

            override fun ime(): ImeHandler? {
                return null
            }

            override fun window(): Window? {
                return null
            }

            override fun logs(): Logs? {
                return null
            }
        }
    }

    private class JSoupWebElement(private val element: Element) : WebElement, FindsByCssSelector, FindsByXPath,
        FindsById, FindsByName, FindsByClassName, FindsByTagName, FindsByLinkText {
        override fun click() {}
        override fun submit() {}
        override fun sendKeys(vararg keysToSend: CharSequence) {}
        override fun clear() {
            element.`val`("")
        }

        override fun getTagName(): String {
            return element.tagName()
        }

        override fun getAttribute(name: String): String? {
            return if (element.hasAttr(name)) element.attr(name) else null
        }

        override fun isSelected(): Boolean {
            return false
        }

        override fun isEnabled(): Boolean {
            return true
        }

        override fun getText(): String {
            return element.text()
        }

        override fun findElements(by: By): List<WebElement> {
            return by.findElements(this)
        }

        override fun findElement(by: By): WebElement {
            return by.findElement(this)
        }

        override fun isDisplayed(): Boolean {
            return false
        }

        override fun getLocation(): Point? {
            return null
        }

        override fun getSize(): Dimension? {
            return null
        }

        override fun getRect(): Rectangle? {
            return null
        }

        override fun getCssValue(propertyName: String): String {
            return ""
        }

        @Throws(WebDriverException::class)
        override fun <X> getScreenshotAs(target: OutputType<X>): X? {
            return null
        }

        override fun findElementByClassName(using: String): WebElement {
            return className.setSelector(using).findElement(this)
        }

        override fun findElementsByClassName(using: String): List<WebElement> {
            return className.setSelector(using).findElements(this)
        }

        override fun findElementById(using: String): WebElement {
            return id.setSelector(using).findElement(this)
        }

        override fun findElementsById(using: String): List<WebElement> {
            return id.setSelector(using).findElements(this)
        }

        override fun findElementByLinkText(using: String): WebElement {
            return linkText.setSelector(using).findElement(this)
        }

        override fun findElementsByLinkText(using: String): List<WebElement> {
            return linkText.setSelector(using).findElements(this)
        }

        override fun findElementByPartialLinkText(using: String): WebElement {
            return partialLinkText.setSelector(using).findElement(this)
        }

        override fun findElementsByPartialLinkText(using: String): List<WebElement> {
            return partialLinkText.setSelector(using).findElements(this)
        }

        override fun findElementByName(using: String): WebElement {
            return name.setSelector(using).findElement(this)
        }

        override fun findElementsByName(using: String): List<WebElement> {
            return name.setSelector(using).findElements(this)
        }

        override fun findElementByTagName(using: String): WebElement {
            return Companion.tagName.setSelector(using).findElement(this)
        }

        override fun findElementsByTagName(using: String): List<WebElement> {
            return Companion.tagName.setSelector(using).findElements(this)
        }

        override fun findElementByCssSelector(using: String): WebElement {
            return cssSelector.setSelector(using).findElement(this)
        }

        override fun findElementsByCssSelector(using: String): List<WebElement> {
            return cssSelector.setSelector(using).findElements(this)
        }

        override fun findElementByXPath(using: String): WebElement {
            return xPath.setSelector(using).findElement(this)
        }

        override fun findElementsByXPath(using: String): List<WebElement> {
            return xPath.setSelector(using).findElements(this)
        }

        private abstract class EditableBy : By() {
            protected var selector: String? = null
            fun setSelector(selector: String?): By {
                this.selector = selector
                return this
            }

            override fun findElement(context: SearchContext): WebElement? {
                return try {
                    super.findElement(context)
                } catch (ex: NoSuchElementException) {
                    null
                }
            }
        }

        companion object {
            // All Bys for JSoupWebElements
            private val cssSelector: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.allElements.stream()
                        .filter { element: Element -> element.cssSelector() == selector }
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val xPath: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement>? {
                    return null
                }
            }
            private val id: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsByAttributeValueContaining("id", selector)
                        .stream().map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val name: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsByAttributeValue("name", selector).stream()
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val className: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsByClass(selector).stream()
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val tagName: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsByTag(selector).stream()
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val linkText: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsMatchingText(selector).stream()
                        .filter { element: Element -> element.tagName() == "a" }
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
            private val partialLinkText: EditableBy = object : EditableBy() {
                override fun findElements(context: SearchContext): List<WebElement> {
                    return (context as JSoupWebElement).element.getElementsContainingText(selector).stream()
                        .filter { element: Element -> element.tagName() == "a" }
                        .map { element: Element -> JSoupWebElement(element) }
                        .collect(Collectors.toCollection { LinkedList() })
                }
            }
        }
    }

    companion object {
        private const val USER_AGENT = "Mozilla/5.0 (Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0"
    }

    init {
        cookies = HashMap()
        generateOptions()
    }
}